interfaces/seccomp/spec.go

```diff
@@ -0,0 +1,98 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package seccomp
+
+import "github.com/snapcore/snapd/interfaces"
+
+// Specification keeps all the seccomp snippets.
+type Specification struct {
+	// Snippets are indexed by security tag.
+	Snippets     map[string][][]byte
+	securityTags []string
+}
+
+// AddSnippet adds a new seccomp snippet.
+func (spec *Specification) AddSnippet(snippet []byte) error {
+	if len(spec.securityTags) == 0 {
+		return nil
+	}
+	if spec.Snippets == nil {
+		spec.Snippets = make(map[string][][]byte)
+	}
+	for _, tag := range spec.securityTags {
+		spec.Snippets[tag] = append(spec.Snippets[tag], snippet)
+	}
+
+	return nil
+}
+
+// Implementation of methods required by interfaces.Specification
+
+// AddConnectedPlug records seccomp-specific side-effects of having a connected plug.
+func (spec *Specification) AddConnectedPlug(iface interfaces.Interface, plug *interfaces.Plug, slot *interfaces.Slot) error {
+	type definer interface {
+		SecCompConnectedPlug(spec *Specification, plug *interfaces.Plug, slot *interfaces.Slot) error
+	}
+	if iface, ok := iface.(definer); ok {
+		spec.securityTags = plug.SecurityTags()
+		defer func() { spec.securityTags = nil }()
```

Comment:
This is to reset securityTags after the call to interface-supplied function (the "definer") finishes. There is no concurrency involved around this code.